core/src/main/scala/codesearch/core/index/repository/SourcePackage.scala

```diff
@@ -0,0 +1,63 @@
+package codesearch.core.index.repository
+
+import sys.process._
+import java.io.File
+import java.net.{URI, URLEncoder}
+import java.nio.file.{Path, Paths}
+
+import codesearch.core.util.Helper
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+
+trait Extractor {
+  def extract(archive: File, directory: File): Future[File] =
+    Future {
+      Seq("tar", "-xvf", archive.getCanonicalPath, "-C", directory.getCanonicalPath) !!;
+      directory
+    }
+}
+
+trait SourcePackage extends Extractor {
+  def fsPath: Path
+  def url: URI
+  def extentions: Set[String] = Set.empty
+}
+
+case class HackagePackage(
+    name: String,
+    version: String
+) extends SourcePackage {
+  val fsPath: Path = Paths.get(new URI(s"./hackage/$name/$version"))
+  val url: URI     = URI.create(s"https://hackage.haskell.org/package/$name-$version/$name-$version.tar.gz")
+}
+
+case class GemPackage(
+    name: String,
+    version: String
+) extends SourcePackage {
+  val fsPath: Path                     = Paths.get(new URI(s"./gem/$name/$version"))
+  val url: URI                         = URI.create(s"https://rubygems.org/downloads/$name-$version.gem")
+  override def extentions: Set[String] = Helper.langByExt.keySet
```

Comment:
Why is `langByExt` used here and but not in other places?

(NB: I don't know whether it's better to get rid of `langByExt` and inline it everywhere, or do the opposite)

Reply:
I did by analogy with existing functions

Reply:
I will check this logic

Reply:
@neongreen we can create enums with all meta information about language - e.g. extensions

I wanted to do it in https://github.com/aelve/codesearch/issues/71